xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,18 +267,146 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
     return count;
 }
 
+int
+__afr_fix_open_cbk(int ret, call_frame_t *frame, void *opaque)
+{
+    AFR_STACK_DESTROY(frame);
+    return 0;
+}
+
+static int
+__afr_getlk_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
+                int32_t op_ret, int32_t op_errno, struct gf_flock *lock,
+                dict_t *xdata)
+{
+    afr_local_t *local = frame->local;
+    int i = (long)cookie;
+    int ret = -1;
+    int32_t strict_locks = -1;
+
+    local->replies[i].valid = 1;
+    local->replies[i].op_ret = op_ret;
+    local->replies[i].op_errno = op_errno;
+    if (op_ret != 0) {
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, AFR_MSG_LK_HEAL_DOM,
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    } else {
+        local->cont.lk.getlk_rsp1[i] = *lock;
+    }
+
+    if (!local->strict_locks) {
+        ret = dict_get_int32(xdata, "strict-locks", &strict_locks);
+        if (ret == 0 && strict_locks)
+            local->strict_locks = _gf_true;
+    }
+
+    syncbarrier_wake(&local->barrier);
+    return 0;
+}
+
+int
+__afr_fix_open(void *opaque)
+{
+    call_frame_t *frame = NULL;
+    xlator_t *this = NULL;
+    afr_private_t *priv = NULL;
+    afr_local_t *local = NULL;
+    dict_t *xdata = NULL;
+    gf_boolean_t locks_held = _gf_false;
+    unsigned char *success_replies = NULL;
+    unsigned char *need_open = NULL;
+    int i = 0;
+    int need_open_count = 0;
+    struct gf_flock flock = {
+        0,
+    };
+
+    frame = (call_frame_t *)opaque;
+    local = frame->local;
+    this = frame->this;
+    priv = this->private;
+
+    local->cont.lk.getlk_rsp1 = GF_CALLOC(sizeof(*local->cont.lk.getlk_rsp1),
+                                          priv->child_count, gf_afr_mt_gf_lock);
+
+    flock.l_type = F_WRLCK;
+    afr_set_lk_owner(frame, this, frame->root);
+    flock.l_owner.len = frame->root->lk_owner.len;
+    if (flock.l_owner.len > 0) {
+        memcpy(flock.l_owner.data, frame->root->lk_owner.data,
+               flock.l_owner.len);
```

Comment:
What is the value of frame->root->lk_owner.len? Since this is a frame you created, is it always 0?

Reply:
Yes, it will be 0 since the frame is created by us.